File: aavt/src/main/java/com/wuwang/aavt/av/CameraRecorder.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.wuwang.aavt.av;



import com.wuwang.aavt.media.SurfaceEncoder;
import com.wuwang.aavt.media.SoundRecorder;
import com.wuwang.aavt.media.SurfaceShower;
import com.wuwang.aavt.media.VideoSurfaceProcessor;


import com.wyz.common.api.IHardStore;
import com.wyz.common.api.Renderer;
import com.wyz.common.core.camera.CameraProvider;
import com.wyz.common.core.mux.StrengthenMp4MuxStore;

/**
 * CameraRecorder2 相机预览及录制工具类
 */
public class CameraRecorder {

    private VideoSurfaceProcessor mTextureProcessor;

    private IHardStore mMuxer;

    private SurfaceShower mShower;
    private SurfaceEncoder mSurfaceStore;
    private SoundRecorder mSoundRecord;

    public CameraRecorder() {
        //用于视频混流和存储
        mMuxer = new StrengthenMp4MuxStore(true); // 存储消费中进行编码的工具
        //用于预览图像
        mShower = new SurfaceShower(); // 展示消费
        mShower.setOutputSize(720, 1280); // 设置输出大小 可以后期调整为自定义获取机器的大小
        //用于编码图像
        mSurfaceStore = new SurfaceEncoder(); // 保存消费
        mSurfaceStore.setStore(mMuxer); // 视频设置编码工具
        //用于音频
        mSoundRecord = new SoundRecorder(mMuxer);// 音频设置编码工具
        //用于处理视频图像
        VideoSurfaceProcessor videoSurfaceProcessor = new VideoSurfaceProcessor();
        videoSurfaceProcessor.setTextureProvider(new CameraProvider());// 摄像头提供者
        videoSurfaceProcessor.addObserver(mShower);
        videoSurfaceProcessor.addObserver(mSurfaceStore);
        mTextureProcessor = videoSurfaceProcessor;
    }

    public void setRenderer(Renderer renderer) {
        mTextureProcessor.setRenderer(renderer);
    }

    /**
     * 设置预览对象，必须是{@link android.view.Surface}、{@link android.graphics.SurfaceTexture}或者
     * {@link android.view.TextureView}
     *
     * @param surface 预览对象
     */
    public void setSurface(Object surface) {
        mShower.setSurface(surface);
    }

    /**
     * 设置录制的输出路径
     *
     * @param path 输出路径
     */
    public void setOutputPath(String path) {
        mMuxer.setOutputPath(path);
    }

    /**
     * 设置预览大小
     *
     * @param width  预览区域宽度
     * @param height 预览区域高度
     */
    public void setPreviewSize(int width, int height) {
        mShower.setOutputSize(width, height);
    }

    /**
     * 打开数据源
     */
    public void open() {
        mTextureProcessor.start();
    }

    /**
     * 关闭数据源
     */
    public void close() {
        mTextureProcessor.stop();
        stopRecord();
    }

    /**
     * 打开预览
     */
    public void startPreview() {
        mShower.open();
    }

    /**
     * 关闭预览
     */
    public void stopPreview() {
        mShower.close();
    }

    /**
     * 开始录制
     */
    public void startRecord() {
        mSurfaceStore.open();
        mSoundRecord.start();
    }

    /**
     * 关闭录制
     */
    public void stopRecord() {
        mSoundRecord.stop();
        mSurfaceStore.close();
        try {
            mMuxer.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

}

File: common/src/main/java/com/wyz/common/CameraRecorder.kt
package com.wyz.common

import android.graphics.Bitmap
import android.opengl.EGLSurface
import android.opengl.GLES20
import android.os.Environment
import android.util.Log
import com.wyz.common.api.FrameDrawedListener
import com.wyz.common.api.IHardStore
import com.wyz.common.api.Renderer
import com.wyz.common.core.audio.SoundRecorder
import com.wyz.common.core.base.FrameBean
import com.wyz.common.core.mux.Mp4MuxStore
import com.wyz.common.core.processor.SurfaceProcessor
import com.wyz.common.core.texture.consume.FrameEncoder
import com.wyz.common.core.texture.consume.FrameShower
import com.wyz.common.core.texture.provider.CameraProvider
import com.wyz.common.utils.DensityUtils
import java.io.*
import java.nio.ByteBuffer
import java.nio.IntBuffer


class CameraRecorder {
    private val mTextureProcessor: SurfaceProcessor

    private val mMuxer: IHardStore

    private val mShower: FrameShower

    private val mSurfaceStore: FrameEncoder

    private val mSoundRecord: SoundRecorder

    constructor() {
        //用于视频混流和存储
        mMuxer = Mp4MuxStore(true) // 存储消费中进行编码的工具
        //用于预览图像
        mShower = FrameShower() // 展示消费
        mShower.setOutputSize(720, 1280) // 设置输出大小 可以后期调整为自定义获取机器的大小
        mShower.setOnDrawEndListener(object : FrameDrawedListener {
            override fun onDrawEnd(surface: EGLSurface?, bean: FrameBean) {
                if (takePictures) {
                    val width = mShower.getWidth()
                    val height = mShower.getHeight()
                    takePictures = false
                    val allocate = IntBuffer.allocate(width * height)
                    GLES20.glReadPixels(0, 0, width, height, GLES20.GL_RGBA, GLES20.GL_UNSIGNED_BYTE, allocate)
                    Thread(Runnable {
                        val result = IntArray(width * height)
                        //解决方向颠倒问题
                        //解决方向颠倒问题
                        for (i in 0 until height) {
                            for (j in 0 until width) {
                                val pix: Int = allocate.get(i * width + j)
                                val pb = pix shr 16 and 0xff
                                val pr = pix shl 16 and 0x00ff0000
                                val pix1 = pix and -0xff0100 or pr or pb
                                result[(height - i - 1) * width + j] = pix1
                            }
                        }
                        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_4444)
                        bitmap.copyPixelsFromBuffer(IntBuffer.wrap(result))
                        val file = File(Environment.getExternalStorageDirectory().absolutePath, "testPicture_${System.currentTimeMillis()}.jpeg")
                        if (!file.exists()) {
                            try {
                                file.createNewFile()
                            } catch (e: IOException) {
                                e.printStackTrace()
                            }
                        }
                        try {
                            val fos = FileOutputStream(file)
                            val bos = BufferedOutputStream(fos)
                            bitmap.compress(Bitmap.CompressFormat.JPEG, 100, bos);
                            bos.flush();
                            bos.close();
                        } catch (e: FileNotFoundException) {
                            e.printStackTrace()
                        }
                    }).start()
                }
            }
        })
        //用于编码图像
        mSurfaceStore = FrameEncoder(mMuxer) // 保存消费 视频设置编码工具
        //用于音频
        mSoundRecord = SoundRecorder(mMuxer) // 音频设置编码工具
        //用于处理视频图像
        val videoSurfaceProcessor = SurfaceProcessor()
        videoSurfaceProcessor.setTextureProvider(CameraProvider()) // 摄像头提供者
        videoSurfaceProcessor.addObserver(mShower)
        videoSurfaceProcessor.addObserver(mSurfaceStore)
        mTextureProcessor = videoSurfaceProcessor
    }

    fun setRenderer(renderer: Renderer) {
        mTextureProcessor.setRenderer(renderer)
    }

    fun setSurface(surface: Any) {
        mShower.setSurface(surface)
    }

    fun setOutputPath(path: String) {
        mMuxer.setOutputPath(path)
    }

    fun setPreviewSize(width: Int, height: Int) {
        mShower.setOutputSize(width, height)
    }

    fun open() {
        mTextureProcessor.start()
    }

    fun close() {
        mTextureProcessor.stop()
        stopRecord()
    }

    fun startPreview() {
        mShower.open()
    }

    fun stopPreview() {
        mShower.close()
    }

    @Volatile
    var takePictures: Boolean = false
    fun takePictures() {
        takePictures = true;
    }

    fun startRecord() {
        mSurfaceStore.open()
        mSoundRecord.start()
    }

    fun stopRecord() {
        mSoundRecord.stop()
        mSurfaceStore.close()
        try {
            mMuxer.close()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }


    interface Listener {
        fun picture(arrays: ByteArray)
    }

}